fiftyone/core/singletons.py

```diff
@@ -24,19 +24,22 @@ def __new__(metacls, *args, **kwargs):
         return cls
 
     def __call__(cls, name=None, _create=True, *args, **kwargs):
+        instance = cls._instances.pop(name, None)
```

Comment:
@kaixi-wang I believe this integrates your idea from https://github.com/voxel51/fiftyone/pull/2634 in a more holistic way: apparently it is possible in multithread/multiprocess scenarios for `load_dataset()` to encounter a `Dataset` object with `deleted==False` but `name==None`. So, here we detect that and decide to reload the dataset from the database in that case.

Reply:
If somehow instance.name is None here, isn't this just creating an instance with no name

Reply:
If `instance.name` is None, this will remove that instance from the dict, which is good.

`name=None` is possible. That's why `name = instance.name` is done later. `Dataset()` internally generates a unique name when called with no `name` arg.